# Exercise 3

**Question 1**

We want to minimize (over w) the empirical risk. In order to do so, we will apply the stochastic gradient descent algorithm. The latter will look like this :

* We choose a $w_0$ randomly and a nonnegative sequence $(\epsilon_k)_k$ which satisfies some properties (we will detail later).

* for k : 1 $\rightarrow$ end :
  * $z_{k+1} \sim \mu$ (here $\mu$ is the uniform distribution on the set of observations ${ \left\{z_i = (x_i​,y_i​) \right\} }_{1⩽i⩽n}$ )

  * Calculate the gradient $\nabla_w J(w, z)$

  * $w_{k+1} = w_k - \epsilon_k \nabla_w J(w, z) $

* end.




Therefore, we need to compute the gradient of J with respect to w for the stochastic gradient descent. This gradient is given by : $$ \nabla_w J(w, z) = - 2 x \left( y - w^\mathsf{T} x \right) $$ 

We need to make sure that all the hypothesis of convergence hold. According to the hypothesis of convergence (H3) seen in class, the gradient $ \nabla_w J(w, z) $ must be bounded. It is patent that this condition is not satisfied here. Indeed if we take $ w = tx $ and let $t \to \infty$ the gradient is clearly not bounded. One solution to this problem is to normalize $w$. This yields : $$ \left\| \nabla_w J(w, z) \right\| \leqslant 2 \left( 1 + \| x \| \right) \, \| x \| $$
Since the number of observation is finite, the norm of $x$ is bounded, and therefore the gradient is bounded. This normalization of w does not affect the hyperplane because the norm of $w$ has no influence on the hyperplane. We will add this step (normalization of w) to our algorithm. $J$ satisfies the rest of the hypothesis of convergence.
The last step is to choose a sequence $(\epsilon_k)_k$ that satisfies the last hypothesis of convergence (H5) of the course, which is : 
$$ \sum{\epsilon_k}  = + \infty, \qquad \sum_{k > 0} \epsilon_k^2 < + \infty \qquad and \qquad  \forall k \qquad \epsilon_k > 0 \qquad $$

In our algorithm we will choose the sequence of inversed integers as the sequnce of $\epsilon_k$. The following algorithm converges toward an optimal solution :

In [ ]:
import numpy as np
import pylab as pl
import matplotlib.pyplot as plt



def stochastic_gradient_descent(Z, w0, steps) : 

	# This function computes and returns a hyperplane w that separates labeled observations contained in Z using a
	# stochastic gradient descent.

	# Z is the list of observations
	# w0 represents the hyperplane used to start the algorithm
	# epsilon	represents the sequence of hypothesis (H5)of the course. Epsilon is used to scale the updates of w

	w = w0
	for k in range(steps) : 
		# draw an observation
		x, y = Z[np.random.randint(len(Z))]
		# compute the gradient
		grad = - 2 * (y - np.dot(w, x)) * x
		# update w
		epsilon = 1/(k+1)
		w = w - epsilon * grad
		# normalization of w 
		w = w/(np.linalg.norm(w))
	return w


**Question 2**

In [ ]:
############ Question 2 #############

# For this question we will create a function generate_sample() that samples a set of observations as asked in the question.

def drawVector():
	# Return a random 2D vector drawn uniformly on the unit sphere
	theta = np.random.uniform(0, 2*np.math.pi)
	return np.array((np.math.cos(theta), np.math.sin(theta)))
 

def generate_sample(n):
	# Function that generates and returns a list of observations z = (x, y).
	# n is the number of observations generated

	w = drawVector()
	print("Real vector w_bar =", w) # will display the real vector w
	sample = []
	for i in range(n):
		r = np.random.rand()
		x = (r ** 0.5) * drawVector() #  x a 2D vector drawn uniformly using the function drawVector()
    
		# Now we compute the label y of x with respect to the side of the hyperplane w the point x is.
		# y takes values in {-1,1}
		if np.dot(w, x) > 0 :      # Recall that w is a hyperplane drawn uniformly on the unit sphere
			y = 1
		else :
			y = -1 
		sample.append([x, y])
	
	return sample


In [ ]:
n = 10 # number of observations

Z = generate_sample(n)
print(" ")
print("Here is a set of observations (of size n=10) that our function generated :")
print(" ")
print(Z)

Real vector w_bar = [0.23374048 0.97229902]
 
Here is a set of observations (of size n=10) that our function generated :
 
[[array([0.2054331 , 0.03086069]), 1], [array([-0.26944663, -0.37525375]), -1], [array([ 0.78858238, -0.05321048]), 1], [array([-0.62028342,  0.17433375]), 1], [array([-0.43766308,  0.69651809]), 1], [array([-0.37308494, -0.75491072]), -1], [array([-0.75696117, -0.41548621]), -1], [array([ 0.27485382, -0.76610157]), -1], [array([-0.5758344,  0.5532276]), 1], [array([-0.18266714, -0.87203057]), -1]]


**Question 3**

In [ ]:
############ Question 3 #############

# Before testing our algorithm, we define a function score() that will compute the score of well predicted observations for a given 
# hyperplane w

def score(Z, w):

	# Z is the list of observations
	# w is an array of float and represents the normal vector of the hyperplane for which we want to know the score

	n = len(Z)
	correct = 0
	for x, y in Z:
		if y*np.dot(w, x) > 0: 
			correct += 1
	score = correct/n # The score is the ratio between the number of correctly predicted observations and the total number of observations.

	return score

In [ ]:
#### Now we can test our algorithm.

# We start by setting the parameters n and steps :

n = 100             # Number of observations
steps = 500         # Number of steps

Z = generate_sample(n) # we generate a sample of observations Z of size n = 100

##### Now we test our algortihm stochastic_gradient_descent on the observations Z

# we start by generating the vector w0 
w0 = drawVector()

w_optimal = stochastic_gradient_descent(Z, w0, steps) 

print(" ")
print("Vector w estimated : w_* =", w_optimal)
print(" ")
print("The score of well predicted observations is score =", score(Z, w_optimal))


Real vector w_bar = [0.61066322 0.79189042]
 
Vector w estimated : w_* = [0.59954426 0.8003416 ]
 
The score of well predicted observations is score = 1.0


As we can see, the vector w_bar with which the data was generated and the vector w_* estimated by the stochastic gradient descent are almost the same. Moreover, the score of well predicted observations is very high. This example illustrates how effective the stochastic gradient descent can be.

**Question 4**

In [ ]:
############ Question 4 #############

# We start by creating a function that noises the observations
def addNoise(Z, sigma):

	# Z is the list of observations
	# sigma is the standard deviation used in the gaussian
	
	for x, y in Z:
		x = x + np.random.normal(np.array([0, 0]), sigma)

In [ ]:
### Now we can perform the optimization again.

# We start by setting the parameters n, sigme and K :
n = 100             # Number of observations
sigma = 0.2         # Standard deviation of the noise
steps = 500             # Number of steps


# We generate our sample of observations
Z = generate_sample(n)
# We add gaussian noise to Z
addNoise(Z, sigma) 

# we generate w0
w0 = drawVector()

w_optimal = stochastic_gradient_descent(Z, w0, steps) 

print(" ")
print("Vector w estimated : w_* =", w_optimal)
print(" ")
print("The score of well predicted observations is score =", score(Z, w_optimal))

Real vector w_bar = [-0.93149074  0.36376503]
 
Vector w estimated : w_* = [-0.88595575  0.46376978]
 
The score of well predicted observations is score = 0.99


As we can see, the result here is nearly as good as the result in question 3. The w_* estimated is really close to the real w_bar and the score of well predicted observations is really high, as in question 3. 
The stochastic gradient descent algorithm seems to handle pretty well the presence of gaussian noise in the observations. Here we chose sigma = 0.2

**Question 5**

In [ ]:
############ Question 5 #############

# First of all, we need to upload the new data Breast_Cancer_data.csv and create our sample of
# observation. This process takes some lines of code.
# The corrector may skip this part and go the next part where we test our algorithm.

import pandas as pd

data = pd.read_csv("/content/Breast_Cancer_data.csv")
df = pd.DataFrame(data)
print(df.columns)

# We first create the vector y of label : we will denote 0 = M and 1 = B
y = []
for i in df["diagnosis"] :
  if (i == "M") : 
    A = 0
    y.append(A)
  else :
    A = 1
    y.append(A)


df = df.dropna(how='all', axis=1) #there is an empty column in the data that we need to drop

# We create the vector x of features
x = []
for j in range(len(df["id"])) :
  B =[]
  for i in df.columns :
    if (i != "diagnosis") :
      B.append(df[i][j])
  x.append(B)

x = np.array(x)

Z=[] 
for i in range(569) :
  Z.append([x[i],y[i]])

# Z is the set of observations of the new data

Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst', 'Unnamed: 32'],
      dtype='object')


In [ ]:
############## Testing our algorithm on Breast_Cancer_data.csv ##############

# In order to test the algorithm stochastic_gradient_descent on the new data, we need to create a new function drawVector31()
# (because the dimension of X here is 569 x 31)

def drawVector31():
  theta = np.random.uniform(0, 2*np.math.pi)
  theta2 = np.random.uniform(0, 2*np.math.pi)
  theta3 = np.random.uniform(0, 2*np.math.pi)
  theta4 = np.random.uniform(0, 2*np.math.pi)
  theta5 = np.random.uniform(0, 2*np.math.pi)
  theta6 = np.random.uniform(0, 2*np.math.pi)
  theta7 = np.random.uniform(0, 2*np.math.pi)
  theta8 = np.random.uniform(0, 2*np.math.pi)
  theta9 = np.random.uniform(0, 2*np.math.pi)
  theta10 = np.random.uniform(0, 2*np.math.pi)
  theta11 = np.random.uniform(0, 2*np.math.pi)
  theta12 = np.random.uniform(0, 2*np.math.pi)
  theta13 = np.random.uniform(0, 2*np.math.pi)
  theta14 = np.random.uniform(0, 2*np.math.pi)
  theta15 = np.random.uniform(0, 2*np.math.pi)
  theta16 = np.random.uniform(0, 2*np.math.pi)
  return np.array((np.math.cos(theta), np.math.sin(theta), np.math.cos(theta2), np.math.sin(theta2), np.math.cos(theta3), np.math.sin(theta3),np.math.cos(theta4), np.math.sin(theta4),np.math.cos(theta5), np.math.sin(theta5),np.math.cos(theta6), np.math.sin(theta6),np.math.cos(theta7), np.math.sin(theta7),np.math.cos(theta8), np.math.sin(theta8),np.math.cos(theta9), np.math.sin(theta9),np.math.cos(theta10), np.math.sin(theta10),np.math.cos(theta11), np.math.sin(theta11),np.math.cos(theta12), np.math.sin(theta12),np.math.cos(theta13), np.math.sin(theta13),np.math.cos(theta14), np.math.sin(theta14),np.math.cos(theta15), np.math.sin(theta15),np.math.cos(theta16)))

##### We set the parameters for our algorithm

steps = 500 # number steps

w0 = drawVector31()

w_optimal = stochastic_gradient_descent(Z, w0, steps)

print(" ")
print("Vector w estimated : w_* =", w_optimal)
print(" ")
print("The score of well predicted observations is score =", score(Z, w_optimal))


 
Vector w estimated : w_* = [9.99999868e-01 1.06954001e-05 1.87497957e-05 6.86304874e-05
 3.12311027e-04 9.44281576e-08 1.02745318e-07 9.37712467e-08
 2.55193182e-08 2.26689929e-07 7.93303411e-08 2.05312490e-07
 1.04983268e-06 1.59106046e-06 1.34388381e-05 6.62923300e-09
 3.86464020e-08 5.59821692e-08 9.47510460e-09 1.94846452e-08
 4.48814884e-09 1.19691392e-05 2.56863294e-05 7.93303411e-05
 3.93701174e-04 1.37283243e-07 3.80340274e-07 4.83330542e-07
 9.04087537e-08 3.32018354e-07 1.09392365e-07]
 
The score of well predicted observations is score = 0.6274165202108963


The score of well predicted observations is 0.63.
I was expecting a higher score. I don't know if it is normal or if something went wrong in my algorithm. 